common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java

```diff
@@ -320,12 +318,25 @@ private ManagedBuffer getSortBasedShuffleBlockData(
       ShuffleIndexInformation shuffleIndexInformation = shuffleIndexCache.get(indexFile);
       ShuffleIndexRecord shuffleIndexRecord = shuffleIndexInformation.getIndex(
         startReduceId, endReduceId);
-      return new FileSegmentManagedBuffer(
-        conf,
-        ExecutorDiskUtils.getFile(executor.localDirs, executor.subDirsPerLocalDir,
+      if (shuffleIndexInformation.isHasDigest()) {
+        File dataFile = ExecutorDiskUtils.getFile(executor.localDirs, executor.subDirsPerLocalDir,
+          "shuffle_" + shuffleId + "_" + mapId + "_0.data");
+        return new DigestFileSegmentManagedBuffer(
+          conf,
+          dataFile,
+          shuffleIndexRecord.getOffset(),
+          shuffleIndexRecord.getLength(),
+          shuffleIndexRecord.getDigest().orElse(DigestUtils.getDigest(
```

Comment:
IIUC, if we calculate digest on the fly for continuous shuffle read, I think we could guarantee the end-to-end consistency.